ion-hash/src/representation.rs

```diff
@@ -120,9 +137,99 @@ fn write_repr_float<D: Update>(value: Option<f64>, hasher: &mut EscapingDigest<'
     }
 }
 
-fn write_repr_string<D: Update>(value: Option<&str>, hasher: &mut EscapingDigest<'_, D>) {
+fn write_repr_symbol<D, S>(value: Option<&S>, hasher: &mut EscapingDigest<'_, D>)
+where
+    D: Update,
+    S: SymbolToken + ?Sized,
+{
     match value {
-        Some(s) => hasher.update(s.as_bytes()),
+        Some(s) => {
+            match s.text() {
+                Some(s) => write_repr_string(Some(s), hasher),
+                None => {
+                    // FIXME: Throw an error if this is not $0
+                    {}
+                }
+            }
+        }
         None => {}
     }
 }
+
+fn write_repr_string<D: Update>(value: Option<&str>, hasher: &mut EscapingDigest<'_, D>) {
+    match value {
+        Some(s) if s.len() > 0 => hasher.update(s.as_bytes()),
+        _ => {}
+    }
+}
+
+fn write_repr_blob<D: Update>(value: Option<&[u8]>, hasher: &mut EscapingDigest<'_, D>) {
+    match value {
+        Some(bytes) if bytes.len() > 0 => hasher.update(bytes),
+        _ => {}
+    }
+}
+
+fn write_repr_struct<D, S, F, E>(
+    value: Option<&S>,
+    hasher: &mut EscapingDigest<'_, D>,
+) -> IonResult<()>
+where
+    D: Update + FixedOutput + Reset + Clone + Default,
+    S: Struct<FieldName = F, Element = E> + ?Sized,
+    F: SymbolToken + ?Sized,
+    E: Element + ?Sized,
+{
+    if let Some(strukt) = value {
+        let mut hashes: Vec<_> = strukt
+            .iter()
+            .map(|(key, value)| struct_field_hash::<D, _, _>(key, value))
+            .collect::<IonResult<_>>()?;
+
+        hashes.sort();
+
+        for hash in hashes {
+            hasher.update(hash);
+        }
+    }
+
+    Ok(())
+}
+
+/// Iterates over a `Struct`, computing the "field hash" of each field
+/// (key-value pair). The field hash is defined in the spec as:
+///
+/// ```text
+/// H(field) -> h(s(fieldname) || s(fieldvalue))
+/// ```
+///
+/// The resulting `Vec` is not sorted (i.e. is in the same order as the field
+/// iterator).
+fn struct_field_hash<D, F, E>(key: &F, value: &E) -> IonResult<Output<D>>
+where
+    D: Update + FixedOutput + Reset + Clone + Default,
+    F: SymbolToken + ?Sized,
+    E: Element + ?Sized,
+{
+    let mut hasher = D::default();
```

Comment:
I notice the `Reset` trait requirement in the signature above. Could `write_repr_struct` create a single instance of `D`, pass it into `struct_field_hash` and then reset it between fields/calls?

Reply:
I think it could.

My old manager used to tell me "make it work, make it fast". I love approaching work like that. So, my overall thinking is that once we have the features fleshed out, we can do some benchmarking and make sure that the 1.0 release is both feature and performance ready.